### PR TITLE
FTS-2266 Added reverse postal code lookup with max distance

### PR DESCRIPTION
**Pull request by dlfajen**
I found this existing library (https://github.com/AReallyGoodName/OfflineReverseGeocode) that:
1. Reads a GeoNames place names file (http://download.geonames.org/export/dump/)
2. Converts the lat/lon into a 3D point
3. Stores the 3D point into a k-d tree (https://en.wikipedia.org/wiki/K-d_tree).
This allows really fast lookups from lat/lon to a GeoNames place.

I added onto this library by:
1. Also reading a GeoNames postal code file (https://download.geonames.org/export/zip/)
2. Allowing the ability to lookup the postal code given a lat/lon with an optional maximum distance

This library will be used by the reference-data-sdk project to create a "ReversePostalCodeService" which will be used within pipeline-uphoria-processors to lookup the postal code after calculating the mobile location (lat/lon). The "allCountries.txt" file will be loaded into memory (takes ~ 3 seconds locally) within  the pipeline-uphoria-processors server once and then used for lookups until the server is restarted. I am still finalizing the code for the other projects.

### CODE REVIEW

File: src/main/java/geocode/PostalCode.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    PostalCode(Double latitude, Double longitude) {
+        this.placeName = "Search";
+        this.latitude = latitude;
+        this.longitude = longitude;
+        setPoint();
+    }
+
+    private void setPoint() {
```
Comment (dlfajen) [inline src/main/java/geocode/PostalCode.java:89]: This is how the 3D point is created

File: src/main/java/geocode/kdtree/KDTree.java [hunk 2]
```diff
@@ -44,7 +44,22 @@ public KDTree( List<T> items ) {
...
+        return findNearest(search, null);
+    }
+
+    /**
+     * Finds nearest point within a max distance (in km)
+     */
+    public T findNearest(T search, Double maxDistance) {
+        T nearest = findNearest(root, search, 0).location;
+        if (null != maxDistance) {
+            int earthRadiusInKm = 6371;
```
Comment (jameswyly) [inline src/main/java/geocode/kdtree/KDTree.java]: Constant values.... what if the size of the earth changes?!

File: src/main/java/geocode/kdtree/KDTree.java [hunk 3]
```diff
@@ -44,7 +44,22 @@ public KDTree( List<T> items ) {
...
+            double distanceInKm = Math.sqrt(nearest.squaredDistance(search)) * earthRadiusInKm;
+            if (distanceInKm > maxDistance) {
+                nearest = null;
```
Comment (jameswyly) [inline src/main/java/geocode/kdtree/KDTree.java:64]: Is this functionality just for part of our business requirements?

File: src/test/java/geocode/ReversePostalCodeTest.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        ReversePostalCode reversePostalCode = new ReversePostalCode(fileStream);
+        assertThat(reversePostalCode.nearestPostalCode(39.0955, -94.5844, null).postalCode,
+                is("64121"));
+    }
+
+    @Test
+    public void givenLocationInGreenlandWithMaxDistance2_findNearestPostalCode_shouldReturnNoPostalCode() throws IOException {
+        InputStream fileStream = ClassLoader.getSystemResourceAsStream("US.txt");
+        ReversePostalCode reversePostalCode = new ReversePostalCode(fileStream);
+        assertNull(reversePostalCode.nearestPostalCode(78.695697, -41.337372, 2D));
```
Comment (jameswyly) [inline src/test/java/geocode/ReversePostalCodeTest.java:37]: I was going to suggest pretty much this exact test and then I scrolled down

Comment (jameswyly) [reply src/test/java/geocode/ReversePostalCodeTest.java:37]: Should there also be a positive test to ensure that the `findNearestPostalCode` can return a non-null value if you give a non-null distance boundary?

**Review by nicolaskline — APPROVED** [review]

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (jameswyly) [reply src/main/java/geocode/PostalCode.java:89]: I presume this is some kind of well-established vector algorithm... but I couldn't find the source with a cursory search. What's the objective of this point conversion?

Comment (jameswyly) [reply src/main/java/geocode/PostalCode.java:89]: Ah is it just trying to normalize our 2d points into 3d vectors?

File: src/main/java/geocode/kdtree/KDTree.java [hunk 5]
```diff
@@ -44,7 +44,22 @@ public KDTree( List<T> items ) {
...
+    }
+
+    /**
+     * Finds nearest point within a max distance (in km)
+     */
+    public T findNearest(T search, Double maxDistance) {
+        T nearest = findNearest(root, search, 0).location;
+        if (null != maxDistance) {
+            int earthRadiusInKm = 6371;
+            double distanceInKm = Math.sqrt(nearest.squaredDistance(search)) * earthRadiusInKm;
```
Comment (jameswyly) [inline src/main/java/geocode/kdtree/KDTree.java]: This _looks_ really confusing without any references to normalization... feels like we might need some extra documentation throughout this to keep it readable and maintainable in the long run.

File: pom.xml [hunk 6]
```diff
@@ -7,7 +7,7 @@
     <groupId>OfflineReverseGeocode</groupId>
```
Comment (jameswyly) [inline pom.xml]: I think if we're pulling this in and maintaining proprietary pieces on top of it, it should become part of our own namespace

File: pom.xml [hunk 7]
```diff
@@ -7,7 +7,7 @@
...
     <artifactId>OfflineReverseGeocode</artifactId>
     <packaging>jar</packaging>
-    <version>1.0-SNAPSHOT</version>
+    <version>1.1-SNAPSHOT</version>
     <name>Offline Reverse Geocoding Java library</name>
     <url>http://github.com/AReallyGoodName/OfflineReverseGeocode</url>
```
Comment (jameswyly) [inline pom.xml]: I'm not sure what the protocol for this is, but we might want to update this to reference our own repository. 

Might also be some scm configurations needed to work with our maven release and jenkins structure. 

And also we might want to bring in the fts parent pom for things like checkstyle

File: src/main/java/geocode/kdtree/KDTree.java (hunk 2 above)
Comment (jameswyly) [reply src/main/java/geocode/kdtree/KDTree.java]: Note, software may become inconsistent after a million years

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (dlfajen) [reply src/main/java/geocode/PostalCode.java:89]: This part of my code is identical to the GeoName.java file, which is already in the library. I couldn't find this exact math anywhere either so I was hoping that @caleboverman can look it over and we can tweak it if we want. I found some distance calculations that were somewhat similar (https://www.movable-type.co.uk/scripts/latlong.html).

File: src/main/java/geocode/kdtree/KDTree.java (hunk 2 above)
Comment (dlfajen) [reply src/main/java/geocode/kdtree/KDTree.java]: 😆 I'll pull this out to a string constant because I use it in my unit tests too.

File: src/main/java/geocode/kdtree/KDTree.java (hunk 3 above)
Comment (dlfajen) [reply src/main/java/geocode/kdtree/KDTree.java:64]: Yes, we didn't want it to return the nearest location if it was over some maximum distance away. It should return no postal code instead.

File: src/test/java/geocode/ReversePostalCodeTest.java (hunk 4 above)
Comment (dlfajen) [reply src/test/java/geocode/ReversePostalCodeTest.java:37]: Yep, I'll add it.

File: src/main/java/geocode/kdtree/KDTree.java (hunk 5 above)
Comment (dlfajen) [reply src/main/java/geocode/kdtree/KDTree.java]: I'll add more documentation.

File: pom.xml (hunk 6 above)
Comment (dlfajen) [reply pom.xml]: Yeah I was wondering how we should handle that. I'll check and see how that hammock one is done.

File: pom.xml (hunk 7 above)
Comment (dlfajen) [reply pom.xml]: Yeah, I wasn't sure either. I know we did something similar for hammock, so I'll see how we handled it there.

File: src/main/java/geocode/kdtree/KDTree.java (hunk 2 above)
Comment (caleboverman) [reply src/main/java/geocode/kdtree/KDTree.java]: WGS-84 defines the radius as 6378.137km
https://en.wikipedia.org/wiki/World_Geodetic_System

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (caleboverman) [reply src/main/java/geocode/PostalCode.java:89]: My java is a little rusty, but it looks like they're using the `squaredDistance` to find the nearest postal code centroid? The haversine method is the preferred geo distance calculation (linked by @dlfajen) and is what we use in the `distance_to_venue` derived attribute.


There's also the distinct possibility that the nearest centroid is not the actual postal code that your lat/lon indicate. A better solution would be using a true reverse geocode API that relies on shapefiles to determine a postal code. That might be a future iteration...

File: src/main/java/geocode/kdtree/KDTree.java (hunk 2 above)
Comment (dlfajen) [reply src/main/java/geocode/kdtree/KDTree.java]: That's the radius at the equator. The 6371 value I'm using is the mean radius (https://en.wikipedia.org/wiki/Earth_radius). Do you think I should switch it?

Comment (caleboverman) [reply src/main/java/geocode/kdtree/KDTree.java]: Mean radius is fine 👍

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (dlfajen) [reply src/main/java/geocode/PostalCode.java:89]: Yes, that's correct. We've already discussed that the FTS lat/lon returns a postal code of 64121 instead of 64108, but we felt like that was probably close enough.

Comment (dlfajen) [reply src/main/java/geocode/PostalCode.java:89]: @caleboverman  - could you point me to the distance_to_venue code? Maybe I can figure out how to implement that here instead.

Comment (caleboverman) [reply src/main/java/geocode/PostalCode.java:89]: https://github.com/sporting-innovations/datasci-enriched-attr/blob/61204ecd5d05d26379ae3ac15815d51ddba3db33/distance_to_venue/distance_to_venue.py#L12

Also:
https://rosettacode.org/wiki/Haversine_formula#Java

File: pom.xml (hunk 6 above)
Comment (dlfajen) [reply pom.xml]: I checked a bunch of other repositories that we've forked to sportinginnovations, but they all left the groupId the same as it was before.

File: pom.xml (hunk 7 above)
Comment (dlfajen) [reply pom.xml]: I added the global_pom. I checked the other forked repositories, but we didn't change the `<url>` value for any of those. I do still need to get the SCM stuff setup for building.

File: pom.xml (hunk 6 above)
Comment (jameswyly) [reply pom.xml]: that may be true, but I'm not sure we care to expose or rely upon the original source naming like that. In fact I think it would be strictly incorrect to use the _exact same_ package name and group information, as it is not the same and theoretically could have versioning conflicts with the real library.  At the very least we should add -FTS or some equivalent... I'd think we should just translate the whole thing over

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (jameswyly) [reply src/main/java/geocode/PostalCode.java:89]: Yeah this doesn't look like the same math, right? I guess I didn't put together how these "point" representations are actually used in every location to see if it maps to a formula we understand... but this is the main usage I see:

point 1 is just cos(lat) * cos(long)
point 2 is cos(lat) * sin (long)
point 3 is sin(lat)

Then those raw values are stored, and when we ask for the `squareDistance` between two "3d points" ([x1,y1,z1], [x2,y2,z2]) it does this:

(x1-x2)^2 + (y1-y2)^2 + (z1-z2)^2

So that expands to roughly (for coordinates [l1,lo1],[l2,lo2]):
((cost(l1) * cos(lo1)) - (cos(l2) * cos(lo2)))^2 + ((cos(l1) * sin (lo1)) - (cos(l2) * sin (lo2)))^2 + (sin(l1) - sin(l2))^2

I'm not sure if that expansion is simplifiable to any of the given formulas above...
Mostly, instead these look like normalization functions for vectors iirc ... I'm still just not really understanding where the mapping from the lat/long to 3d "point" representation comes from. The mapping of a 2d point to a 3d represent

Comment (jameswyly) [reply src/main/java/geocode/PostalCode.java:89]: Oh here we go, lat/long mapped to geodesy vectors : https://www.movable-type.co.uk/scripts/latlong-vectors.html

then it just does normal vector distance calculations

Comment (jameswyly) [reply src/main/java/geocode/PostalCode.java:89]: Is this distance calculation implemented properly though? My trigonometric function conversions are a big rusty... https://www.movable-type.co.uk/scripts/latlong-vectors.html#distance

> d = R · atan2(|a×b|, a·b)
> where R is the radius of the earth (normally taken as 6371 km).

Comment (dlfajen) [reply src/main/java/geocode/PostalCode.java:89]: Hmm, yeah... I'm not finding where I found that distance equation. I'll change it to match what you've posted because that link is exactly what this other code is doing.

Comment (dlfajen) [reply src/main/java/geocode/PostalCode.java:89]: So I changed the distance in the max distance check to use the haversine distance instead. I'm wondering if I should do anything with the other distance checks that are in the KD tree lookups. I think he coded it like this as more of a quick approximation.

Comment (caleboverman) [reply src/main/java/geocode/PostalCode.java:89]: Yeah my impression was it used a quick check to narrow the results. So it’s probably ok to only use haversine once the search space has been narrowed.

File: pom.xml (hunk 6 above)
Comment (dlfajen) [reply pom.xml]: Looked around at standards for forking repositories. It was suggested that changing the groupId, name, and description are a good idea, so I've updated those three items. I picked `com.fanthreesixty.libraries` as the groupId - let me know if something else would be better.

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (jameswyly) [reply src/main/java/geocode/PostalCode.java:89]: Is there a variation of the haversine algorithm for vectors? The (haversine) links above always just use coordinates directly so if you wanted to use that you could skip the vectorization right? 

Otherwise, given the converted vectors you can do the more simple calculations based on the distance between two vectors on a sphere (given the radius) which is the function I quoted above with arctan and the cross product of the vectors.

**Comment by dlfajen** [comment]
I still need to add README and CONTRIBUTING files too, once I figure all that out.

**Comment by jameswyly** [comment]
This is my most up-to-date CONTRIBUTING.md for reference if you want https://github.com/sporting-innovations/pipeline-uphoria-processors/blob/master/CONTRIBUTING.md

This was the only library-ish README that I've made so far... most of them have been for applications which require a lot of configurations. https://github.com/sporting-innovations/bigquery-utils/blob/master/README.md  ....  for this project is the input file configurable? 

And the CHANGELOG is straightforward enough, just need to copy the headers. https://github.com/sporting-innovations/bigquery-utils/blob/master/CHANGELOG.md

It might make sense to use the original forked version with all of these metadata pieces added as maybe version 0.0.1, then you can add all of your updates here as part of 1.0.0 release.... though I guess I don't know if you can do that reasonably with the direct fork. If not, since you're changing the namespace I think your first release here can still be version 1.0.0

File: src/main/java/geocode/PostalCode.java (hunk 1 above)
Comment (dlfajen) [reply src/main/java/geocode/PostalCode.java:89]: Switched to use the distance calculation using arctan(cross product, dot product). Also added a junit test for it.

**Comment by dlfajen** [comment]
So I didn't realize that forked repositories can't be private, but @jbkc85 pointed out that this sporting-innovations/OfflineReverseGeocode repository that I created from a public repository is now also public. Do we think that will be an issue? There isn't anything in the code that is specific to FTS, it's all just math that you can get off the internet. The only thing I can see that I might want to change is the unit test specifying the FTS location. I could call it Kansas City instead.

**Comment by nicolaskline** [comment]
@dlfajen I have no issues with it being a public repository. In fact I still suggest you PR back to the original project.

**Comment by dlfajen** [comment]
Thanks! I plan on doing that once we get it all finalized here.

**Comment by dlfajen** [comment]
Merging this to master to work on getting a jar file built. I'll still work on any changes that anyone suggests.

File: src/main/java/geocode/GeoName.java [hunk 8]
```diff
@@ -92,15 +95,13 @@ protected double axisSquaredDistance(GeoName other, int axis) {
...
-        double dLat = toRadians(other.latitude - this.latitude);
-        double dLon = toRadians(other.longitude - this.longitude);
-        double lat1 = toRadians(this.latitude);
-        double lat2 = toRadians(other.latitude);
-
-        double a = pow(sin(dLat / 2),2) + pow(sin(dLon / 2),2) * cos(lat1) * cos(lat2);
-        return 2 * asin(sqrt(a)) * radius;
+    protected double distance(GeoName other, double radius) {
+        // distance = atan2(cross product, dot product) * radius
+        return atan2(cross(other), dot(other)) * radius;
```
Comment (jameswyly) [inline src/main/java/geocode/GeoName.java:104]: This looks good, as do the cross and dot product functions 👍 

Took me a second to figure out where the magnitude calculation came from but it was right too it looks like
